lib/auth/grpcserver.go

```diff
@@ -1506,6 +1506,34 @@ func (g *GRPCServer) UpsertKubeService(ctx context.Context, req *proto.UpsertKub
 	return new(empty.Empty), nil
 }
 
+// UpsertKubeServer adds a kubernetes service
+func (g *GRPCServer) UpsertKubeServer(ctx context.Context, req *proto.UpsertKubeServiceRequest) (*types.KeepAlive, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	server := req.GetServer()
+	// If Addr in the server is localhost, replace it with the address we see
+	// from our end.
+	//
+	// Services that listen on "0.0.0.0:12345" will put that exact address in
+	// the server.Addr field. It's not useful for other services that want to
+	// connect to it (like a proxy). Remote address of the gRPC connection is
+	// the closest thing we have to a public IP for the service.
+	clientAddr, ok := ctx.Value(ContextClientAddr).(net.Addr)
+	if !ok {
+		return nil, status.Errorf(codes.FailedPrecondition, "bug: client address not found in request context")
+	}
+	server.SetAddr(utils.ReplaceLocalhost(server.GetAddr(), clientAddr.String()))
```

Comment:
Kube service doesn't have a `public_addr` you can use here?

Reply:
Unfortunately in `kubernetes_service` the `public_addr` field is optional. `GetAddr` will return the `public_addr` if it is set, otherwise it returns `listen_addr` which is most likely `0.0.0.0`. If this addr isn't replaced then the proxy will be unable to reach out as the address to dial is set incorrectly.
 
Note this seems to only be required in a narrow field of configuration, `kubernetes_service` is running standalone, connecting directly to auth server, and `public_addr` is not set. The other scenarios I tested worked fine without this logic. Making `public_addr` required when not connecting through a proxy would allow us to get rid of this check from what I can tell.